mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -30,50 +33,51 @@ import kotlinx.coroutines.launch
  */
 private const val debugOutput = false
 
-internal fun mosaicNodes(content: @Composable () -> Unit): MosaicNode {
-	val clock = BroadcastFrameClock()
-	val job = Job()
-	val composeContext = clock + job
-
-	val rootNode = createRootNode()
-	val recomposer = Recomposer(composeContext)
-	val composition = Composition(MosaicNodeApplier(rootNode), recomposer)
-
-	composition.setContent(content)
-
-	job.cancel()
-	composition.dispose()
-
-	return rootNode
-}
-
-public fun renderMosaic(content: @Composable () -> Unit): String {
-	val rootNode = mosaicNodes(content)
-	val render = AnsiRendering().render(rootNode)
-	return render.toString()
-}
-
 public suspend fun runMosaic(content: @Composable () -> Unit) {
 	coroutineScope {
-		val mosaicComposition = MosaicComposition(coroutineScope = this, onRender = ::platformDisplay)
+		val mosaicComposition = RenderingMosaicComposition(
+			coroutineScope = this,
+			onRender = ::platformDisplay,
+		)
 		mosaicComposition.setContent(content)
 		mosaicComposition.awaitComplete()
 	}
 }
 
-internal class MosaicComposition(
+public fun renderMosaic(content: @Composable () -> Unit): String {
+	var singleRender: String? = null
+	val mosaicComposition = RenderingMosaicComposition(
+		coroutineScope = CoroutineScope(EmptyCoroutineContext),
+		onRender = { render ->
+			if (singleRender == null) {
+				singleRender = render.toString()
+			}
+		},
+	)
+	mosaicComposition.setContent(content)
+	mosaicComposition.cancel()
+	return checkNotNull(singleRender) {
+		"It shouldn't happen. We are waiting for synchronous render after calling setContent."
+	}
```

Comment:
```suggestion
	return singleRender!!
```
No need to waste bytes on branches that should be impossible